issue-4023: partial overwrite of metadata holes results in loss of hole birth info

bprotopopov opened the pull request:
When reading Ln metadata holes with non-zero birth epochs, instead of initializing the dbuf data with zeros, initialize with Ln-1 hole block pointers of same dmu type and birth epoch as Ln hole, and of
proper level and logical size, in order to make sure the new (Ln-1) holes inherit correct hole info.

This is important when generating incremental zfs send streams with hole_birth feature active, where
the proper hole birth is required in order to generate proper FREE records for the new holes.

Regression tested with ztest and verified proper assignment of hole birth epochs with zdb and by parsing the resulting send stream.

behlendorf commented:
I haven't had a chance to look closely at the patch but I did notice it causes a failure in xfstests 013.  Stack below.

```
[ 1542.828564] VERIFY(buf[i] == 0) failed
[ 1542.831199] PANIC at dbuf.c:550:dbuf_verify()
[ 1542.833949] Showing stack for process 627
[ 1542.836451] CPU: 1 PID: 627 Comm: fsstress Tainted: PF          O--------------   3.10.0-229.el7.x86_64 #1
[ 1542.842295] Hardware name: Xen HVM domU, BIOS 4.2.amazon 05/06/2015
[ 1542.845819]  ffffffffa04710d4 000000009d5dea6d ffff88034c7e7728 ffffffff81604b0a
[ 1542.850366]  ffff88034c7e7738 ffffffffa025e3f4 ffff88034c7e78c0 ffffffffa025e4bf
[ 1542.854827]  0000000000000000 ffff880300000028 ffff88034c7e78d0 ffff88034c7e7870
[ 1542.858896] Call Trace:
[ 1542.860416]  [<ffffffff81604b0a>] dump_stack+0x19/0x1b
[ 1542.863445]  [<ffffffffa025e3f4>] spl_dumpstack+0x44/0x50 [spl]
[ 1542.866381]  [<ffffffffa025e4bf>] spl_panic+0xbf/0xf0 [spl]
[ 1542.869398]  [<ffffffff816165f1>] ? ftrace_call+0x5/0x2f
[ 1542.872030]  [<ffffffff816165f1>] ? ftrace_call+0x5/0x2f
[ 1542.874992]  [<ffffffffa036671e>] dbuf_verify+0x59e/0x750 [zfs]
[ 1542.878330]  [<ffffffffa0366a5d>] dbuf_rele_and_unlock+0x4d/0x4a0 [zfs]
[ 1542.881528]  [<ffffffff816165f1>] ? ftrace_call+0x5/0x2f
[ 1542.884659]  [<ffffffffa03671cb>] dbuf_rele+0x4b/0x80 [zfs]
[ 1542.887780]  [<ffffffffa03699c7>] __dbuf_hold_impl+0x277/0x900 [zfs]
[ 1542.890856]  [<ffffffffa036a0c3>] dbuf_hold_impl+0x73/0xa0 [zfs]
[ 1542.893868]  [<ffffffffa036a5ce>] dbuf_hold_level+0x2e/0x60 [zfs]
[ 1542.897595]  [<ffffffffa0387779>] dmu_tx_check_ioerr+0x39/0xf0 [zfs]
[ 1542.901255]  [<ffffffffa0387e9e>] dmu_tx_count_write+0x2de/0x760 [zfs]
[ 1542.905182]  [<ffffffff816165f1>] ? ftrace_call+0x5/0x2f
[ 1542.908501]  [<ffffffffa0256c20>] ? spl_kmem_zalloc+0xf0/0x1a0 [spl]
[ 1542.912240]  [<ffffffffa0256c20>] ? spl_kmem_zalloc+0xf0/0x1a0 [spl]
[ 1542.915966]  [<ffffffffa03886da>] dmu_tx_hold_write+0x5a/0x100 [zfs]
[ 1542.919837]  [<ffffffffa042c18b>] zfs_write+0x3eb/0xe80 [zfs]
[ 1542.923451]  [<ffffffff811d3ff1>] ? link_path_walk+0x81/0x8b0
[ 1542.926938]  [<ffffffff816165f1>] ? ftrace_call+0x5/0x2f
[ 1542.930272]  [<ffffffffa03c1bd6>] ? refcount_remove+0x16/0x20 [zfs]
[ 1542.934140]  [<ffffffffa044b3c0>] zpl_write_common_iovec.constprop.8+0x90/0x130 [zfs]
[ 1542.938926]  [<ffffffffa044b56e>] zpl_write+0x7e/0xb0 [zfs]
[ 1542.942113]  [<ffffffff811c6c6d>] vfs_write+0xbd/0x1e0
[ 1542.945229]  [<ffffffff811c76b8>] SyS_write+0x58/0xb0
[ 1542.948320]  [<ffffffff81614a29>] system_call_fastpath+0x16/0x1b
```

bprotopopov commented:
@behlendorf, I am afraid this is unrelated; in my case, there is no panic, and data integrity issue only arises at zfs recv site when one sends zfs send incremental stream (because the latter misses some holes). This bug is similar to the 
Illumos issue #5659: dmu_sync'ed holes should retain birth time

bprotopopov commented:
@behlendorf, I will take a look, thanks for the pointer.

bprotopopov commented:
Yes, let me test some more with ZFS_DEBUG on.

bprotopopov commented:
I believe I have addressed debug-mode issues with the latest commits.

bprotopopov commented:
Incomplete push, working on it

bprotopopov commented:
I am running ztest and with default stack size, I see intermittent asserts, whereas with 64k stack size, I do not see asserts. Is this normal ? 

Also, I looked at the implementation of the range lock in ztest, and I am not sure I understand why it is supposed to work. Perhaps for small object numbers and offsets with large alignments, the range locks become object-scope locks, and they "work". But for unaligned offsets, they seem to easily break.

The reason I ask is that the test I added (ztest_dmu_free_long_range()) requires working range locks. I noticed that the ztest_dmu_prealloc() that also uses range locks and dmu_free_long_range() is disabled both in zfsonlinux and in openzfs code since a very long time ago.

I am not certain how to proceed - should I also disable my test ?

behlendorf commented:
@bprotopopov on what distribution and release?  We do still do see occasional segfaults on CentOS 6 which may be related to stack size.  Although these default stack size is 256K for ztest unless set otherwise so it would be odd that making it smaller helps.  Unless we're just failing to create new threads due to memory being exhausted?

As for the range locks I've actually never looked very carefully at them in ztest, I see it's using it's own version.  But if they're broken we should fix them first so you can add your test case (and we can reenable the other tests).  Actually it would be great if ztest shared the same range lock code so it could also be validated but that might take a fair bit of refactoring.

bprotopopov commented:
@behlendorf, yes I am testing on CentOS 6.7. I am fully rebased to the tip of the master. 

I get intermittent asserts getting NULL out of the thread-specific var in zk_thread_current() that seems to correlate very well with running with default stack. I guess I'll get more run stats. 

Yes, I was planning to look into switching to the kernel range locks implementation bug did not have a chance to do so.

behlendorf commented:
@bprotopopov it's not clear to me how that's possible with the current code.  It has be wondering if we're hitting some grey area in pthreads because I'm fairly certainly we don't see this on CentOS 7 or other more modern Linux distributions with newer versions of pthreads.  I'd be very interested to hear whatever you determine!

As for the range lock it looks like you could use them if you rig up a partial `znode_t` to pass in.

bprotopopov commented:
Rebased against the master. Addressed cstyle issues. Tested with ztest, no issues with standard stack size. It looks like I was slightly out of sync between the zfs and spl code on the test machine. The new MUTEX_NOLOCKDEP macro in dbuf.c forced rebase to master in spl.

kernelOfTruth commented:
@bprotopopov just amend some comments changes to the commit and force-push 

that should do it ;)

behlendorf commented:
Where does this stand?  Was a proper solution agreed upon?

@bprotopopov you can always retrigger a test run by rebasing on master and force updating the branch.   There are definitely still some infrastructure failures, the Github outage yesterday didn't help any.

bprotopopov commented:
@behlendorf, I think this is a working solution. I have developed this concurrent with the solution presented in https://github.com/openzfs/openzfs/pull/46. I believe the latter is a bit off in a few places, I have made my comments to that effect. I think I will summarize those comments again and point to this pull request asking for a review from @pcd1193182 and @ahrens.  

Another issue was the ztest runs asserting intermittently in the txg_sync() thread as it's stack was getting corrupted. I have rebased both SPL and ZFS to the tip of the master, and I do not see those any more. 

Finally, there was an issue with added test case fore dmu_free_long_range() that I felt needed a better implementation for the range locks. I have added such implementation to ztest.

behlendorf commented:
@bprotopopov OK, it sounds like we need to let this settle a bit so we're sure we end up with the same fix as illumos.

The stack issue was addressed by 89666a8e1cfe8691a31fa8ca33c55362e959e2ed, based on your previous observation the stack was slightly increased in user space.

bprotopopov commented:
@behlendorf, sounds good. This pull request is a superset of the one in openzfs, as it includes three additional fixes: one somewhat cosmetic (properly setting the size of L0 holes), and two of a more functional nature - one that adjusts the dn_maxblikid so that no blkids for the new meta-level holes point beyond it in dbuf_write_ready(), and one making sure we are not forgetting the new holes in dnode_increase_indirection(). I have asked the guys (@pcd1193182 and @ahrens) for a review.

pcd1193182 commented on module/zfs/dbuf.c:
```diff
@@ -3074,12 +3131,25 @@ dbuf_write_ready(zio_t *zio, arc_buf_t *buf, void *vdb)
+		 */
+		mutex_enter(&dn->dn_mtx);
+		if (db->db_blkid >
+			dn->dn_phys->dn_maxblkid >> (db->db_level * epbs))
```
continuation lines should be indented by 4 spaces

pcd1193182 commented on module/zfs/dbuf.c:
```diff
@@ -3074,12 +3131,25 @@ dbuf_write_ready(zio_t *zio, arc_buf_t *buf, void *vdb)
 			fill += BP_GET_FILL(ibp);
 		}
+		/*
+		 * it turns out that pre-filling Ln holes in dbuf_read_impl()
```
Capitalize first letter of comment

pcd1193182 commented on module/zfs/dbuf.c:
```diff
@@ -3074,12 +3131,25 @@ dbuf_write_ready(zio_t *zio, arc_buf_t *buf, void *vdb)
+		mutex_enter(&dn->dn_mtx);
+		if (db->db_blkid >
+			dn->dn_phys->dn_maxblkid >> (db->db_level * epbs))
+			dn->dn_phys->dn_maxblkid = ((db->db_blkid + 1) <<
```
I think this violates the intended meaning of dn_maxblkid; I think the idea is that if the file is empty, dn_maxblkid is 0, and otherwise it's the number of the highest blkid of an actual filled in block on that object.  With this new proposal, you could easily end up in situations where dn_maxblkid is much farther along than the last filled-in l0 block.

bprotopopov replied:
Hi, @pcd1193182 
thanks for the review. I agree with all the points but would like to discuss the last one about the maxblkid. The reason I added this tweak was to avoid asserts in dbuf_write_done() (see <<< mark):
```
        } else {
                dnode_t *dn;

                DB_DNODE_ENTER(db);
                dn = DB_DNODE(db);
                ASSERT(list_head(&dr->dt.di.dr_children) == NULL);
                ASSERT3U(db->db.db_size, ==, 1 << dn->dn_phys->dn_indblkshift);
                if (!BP_IS_HOLE(db->db_blkptr)) {
                        ASSERTV(int epbs = dn->dn_phys->dn_indblkshift -
                            SPA_BLKPTRSHIFT);
			ASSERT3U(db->db_blkid, <=,
                            dn->dn_phys->dn_maxblkid >> (db->db_level * epbs));  <<<
			ASSERT3U(BP_GET_LSIZE(db->db_blkptr), ==,
                            db->db.db_size);
			if (!arc_released(db->db_buf))
                                arc_set_callback(db->db_buf, dbuf_do_evict, db);
                }
                DB_DNODE_EXIT(db);
                mutex_destroy(&dr->dt.di.dr_mtx);
                list_destroy(&dr->dt.di.dr_children);
        }
```
Your point about the meaning of the maxblkid makes sense. Do you believe it is OK to have Ln metadata blocks that span logical ranges beyond the maxblkid then ? And if so, should this assert be revised ? 

I understand that what the assert says is as follows: for any metadata block, its block id should be such that the upper bound of the logical data (L0) block id range that it spans does not exceed maxblkid. 

Thinking about this for a moment, I believe that even before my changes, the maxblkid could be larger than the highest data block id. For instance, let's take a a file with a single L0 metadata block (pointed to from the dnode). Say all but the top three of the data blocks are filled, and the last three a holes. In order to satisfy the assert above, the maxblkid should be equal to the block id of the top hole. Which is not the last allocated block.

pcd1193182 commented:
@bprotopopov We've got an updated version of the openzfs patch that addresses most of your comments internally.  It differs from yours in a few ways.  
We copy all the blkptrs when we do dnode_increase_indirection, not just the non-holes and the non-birth time-zero ones; there isn't a significant cost, and it makes the code cleaner.
We set the LSIZE slightly differently, but the result is the same.
Your change to dnode_sync() seems fine.  Doesn't change anything, but it's certainly cleaner.  Not sure why it wasn't that way in the first place!
I'm not sure about the ztest changes; are those ZoL specific?
I left a couple line comments in the code; I'm specifically a little concerned abotu the changes to dbuf_write_ready()

bprotopopov commented:
Hi, @pcd1193182

no, the changes to the ztest are intended to test the new code that deals with holes; the dmu_free_long_range() is a generic mechanism to punch arbitrary holes in objects, including ones that span Ln (n >= 1) block ranges.

In addition, I also implemented range locks based on zfs range locks for ztest; it did not look to me that the old user-space implementation of the range locks in ztest was sufficient for my goals. 

These aren't specific to ZoL at all, to my knowledge.

bprotopopov commented:
Hi, @pcd1193182 

here is the assertion in ztest (with my added test case for dmu_free_lokc_fange()) I hit if I remove the fixup of the maxblkid discussed above:
```
ztest: ../../module/zfs/dbuf.c:3270: Assertion `db->db_blkid <= dn->dn_phys->dn_maxblkid >> (db->db_level * epbs) (0xa3bd0 <= 0x3e8d6)' failed.
/sbin/ztest[0x409093]
/lib64/libpthread.so.0[0x3e6b20f790]
/lib64/libc.so.6(gsignal+0x35)[0x3e6aa32625]
/lib64/libc.so.6(abort+0x175)[0x3e6aa33e05]
/lib64/libc.so.6[0x3e6aa2b74e]
/lib64/libc.so.6(__assert_perror_fail+0x0)[0x3e6aa2b810]
/lib64/libc.so.6[0x3e6aa2b87b]
/lib64/libzpool.so.2(+0x4a52a)[0x7fda9977f52a]
/lib64/libzpool.so.2(+0x38ffe)[0x7fda9976dffe]
/lib64/libzpool.so.2(+0xf9836)[0x7fda9982e836]
/lib64/libzpool.so.2(+0xfa112)[0x7fda9982f112]
/lib64/libzpool.so.2(+0xfa112)[0x7fda9982f112]
/lib64/libzpool.so.2(+0xfa112)[0x7fda9982f112]
/lib64/libzpool.so.2(zio_execute+0xe0)[0x7fda998278f0]
/lib64/libzpool.so.2(+0x2e912)[0x7fda99763912]
/lib64/libzpool.so.2(zk_thread_helper+0x22c)[0x7fda9975fd1c]
/lib64/libpthread.so.0[0x3e6b207a51]
/lib64/libc.so.6(clone+0x6d)[0x3e6aae893d]
child died with signal 6
```

bprotopopov commented:
Updated based on the review comments (all the updates apply to the first commit, the contents of the last two did not change).

pcd1193182 commented:
@bprotopopov a fix is definitely needed for that issue, we've been seeing it internally as well.  However, I don't think that what you're proposing is necessarily the right fix.  @grwilson has a different plan to fix the issue, though I'll admit I'm not entirely clear on what that plan is.  Probably the different approaches should be compared, and we should figure out which one will be the best in the long run.

grwilson commented:
The real problem is that indirect blocks that contain only holes with birth times no longer compress down to 0 bytes when they are written. To fix this I'm having ZIO pipeline zero out indirect blocks that contain only holes so that they will again compress to 0 bytes. I have a preliminary fix for this that I'm still testing.

bprotopopov commented:
Yes, @grwilson and @pcd1193182, I also had a fix for that in zio layer - added function zio_compress_metadata() that worked much like zio_compress_data(). It worked, but this alone did not address the assert. For instance, a partly full L2 metadata block with holes at the end (whether or not those holes point to allocated L1 blocks or not) will definitely have its block id triggering this assert. 

I did not include it this my final version because I thought that the amount of storage taken was trivial, and it was not worth coming up with a workaround.

bprotopopov commented:
@grwilson and @pcd1193182, so, I went back to what I did last time and realized that my approach only dealt with datasets with compression, and therefore was not general enough (was trying to do too many things at once I guess). I modified it to apply more generally, and you are right, it works; the fixup of the maxblkid is not needed.

What I did was in zio_write_bp_init(), set psize to zero if the block with zp->zp_level > 0 is filled entirely with hole bps, and that was sufficient as the code below already deals with this in case of data compression. I will be posting the updated commits shortly.

bprotopopov commented:
@behlendorf, 
quick update on the status of this pull request. @grwilson and @pcd1193182 have made some good points that I have addressed. I checked the openzfs pull requests and could not find the patch set that dealt with the issues discussed above.

bprotopopov commented:
@behlendorf, there is a thought.

It seems like there is a number of concerns regarding the hole_birth feature that have been surfacing, including the above use cases and the dnode reuse scenarios (6370 openzfs). The bugs in the hole_birth support result in data loss when ```zfs send | zfs``` recv or sparse datasets are performed.

Given that the fixes from openzfs/illumos are still in development, I think it might be prudent to temporarily introduce a tunable/zfs module parameter that turns off the zfs send optimizations that are based on the hole_birth feature, so that one could troubleshoot/work around the problems in the field. Once the fixes are in place, it should be easy to remove the tunable (or change its default value).

behlendorf commented:
@bprotopopov @grwilson @pcd1193182 do you have a feeling for how far off an openzfs/illumos fix for this issue is?  If it's going to be months then I think disabling @bprotopopov's idea of disabling these optimization at least is the short term is worth doing.

behlendorf commented:
@bprotopopov the 6370 fix from illumos has been merged in to master and should appear in our next point release.  Could rebase the additional ztest improvements you've made on to master we can look in to getting them included as well.  Do you have any additional concerns?

bprotopopov commented:
Yes, @behlendorf, let me have a quick look

bprotopopov commented:
Hi, @behlendorf, I believe the 6370 addresses a complimentary issue - missed holes in filesystems when dnodes are reused. The pull request above addresses incorrect processing of metadata-level (Ln, n > 0) holes, which is closely related to '6513 partially filled holes lose birth time #46', and is common across both filesystems and zvols.

Yes, I can rebase the ztest portion, such as range locks from the kernel space. However, the ztest changes are not designed to validate the holes transmitted properly. Rather, the changes verify that the fix for the holes does not break any other functionality. Once the 'obvious' issues with partially filled holes were addressed, quite a few less trivial side effects arose that had to be dealt with, and fixes for those side effects needed an additional test case in ztest, and the test case ran into limitations of the user space range locks, as I recall it.

I understand that the preferred way of addressing this is to adopt a fix from openzfs, instead of taking this commit. This is fine with me. In my code base, I have tested the fix above in pre-production, and my production code has been running with the 'hole birth' optimization turned off for a while, so I am not affected by this issue. Other people, however, still are. Frankly, I am very surprised we do not see more reports on this - I guess large holes (minimum of 128*block size) in files or zvols aren't that common.

Like I said, if you prefer, I can rebase this, and you can take the commits with range locks support and/or the ztest changes, they are already separated.

behlendorf commented on include/sys/zfs_rlock.h:
```diff
@@ -34,6 +34,24 @@ extern "C" {
+	kmutex_t z_range_lock;
+	avl_tree_t z_range_avl;
+	boolean_t z_is_zvol;
+} znode_t;
```
I'd consider moving this user space `znode_t` definition in to `zfs_znode.h`.  That should allow you to entirely from the `#ifdef _KERNEL` wrapper from `zfs_rlock.h` which would be nice.  And it would have the benefit of having both versions of `znode_t` be declared in on place.

bprotopopov replied:
Sure I'll look into it

behlendorf commented on module/zfs/zfs_rlock.c:
```diff
@@ -106,10 +106,15 @@ zfs_range_lock_writer(znode_t *zp, rl_t *new)
+	uint64_t end_size;
+#else
+	ASSERT3U(zp->z_is_zvol, ==, B_TRUE);
+#endif
```
It's not clear it me it's worth the bother to add the `z_is_zvol` member if it's only ever used for this assertion.  Also wouldn't it make more sense to set `z_is_zvol = 0` and then initialize `z_size` and `z_blksz` to reasonable values so the `if (!zp->z_is_zvol) { }` wouldn't need to be disabled?

bprotopopov replied:
I believe z_is_zvol is used by the zfs_range_lock_writer() to decide if this is a 'simple' (for zvols) or 'compicated' (for real znodes - files) implementation. This member simplifies things quite a bit in terms of using 'user space znode structure.

behlendorf commented on cmd/ztest/ztest.c:
```diff
@@ -4208,6 +4210,182 @@ ztest_dmu_write_parallel(ztest_ds_t *zd, uint64_t id)
+{
+	objset_t *os = zd->zd_os;
+	uint64_t object = od->od_object;
+	uint64_t size = 1024ULL*1024ULL*1024ULL*1024ULL;
```
(1ULL << 40) might be clearer.

bprotopopov replied:
Makes sense to me, will fix.

behlendorf commented:
@bprotopopov I see.  Yes, in the end we'll want to to adopt the same patch which is applied to openZFS and apply your test cases.  If upstream is working on this already with you I'm happy to wait until it's finalized.

bprotopopov commented:
@tuxoko, here is a rebase against your branch zrl. For some reason, the commits above are in a different order, but the one you are interested in is c4ebefe.

The 78013e0 and 66127cb deal with the original issue, whereas the c7bc5db is a ztest unit test that punches large holes in objects.

tuxoko commented on include/sys/zfs_rlock.h:
```diff
@@ -74,10 +77,15 @@ struct znode;
 	zfs_range_lock_impl(zrl, off, len, type, NULL)
+#else
+#define	zfs_range_lock(zrl, off, len, type) \
+	zfs_range_lock_impl(zrl, off, len, type, NULL)
```
You should just use zfs_range_lock_impl and don't bother with znode at all. Having different prototype of the "same" function for kernel and user is really ugly.

bprotopopov replied:
@tuxoko, do you have a specific suggestion ?

tuxoko commented on include/sys/zfs_znode.h:
```diff
@@ -38,7 +39,24 @@
+	uint64_t z_size;		/* unused, zeroed out */
+	uint64_t z_blksz;		/* unused, zeroed out */
+	uint64_t z_max_blksz;		/* unused, zeroed out */
+	zfs_rlock_t z_range_lock; 
```
If we only need zfs_rlock_t, then just use it directly.

bprotopopov replied:
I think the only way to use zfs_rlock_t directly is to not declare a separate znode structure in user space and to assure the user mode always calls zfs_range_lock_impl() with zp = NULL. I could look into this, but is there a big difference ?

tuxoko replied:
Yes, that's what I meant. Not need to define something if you don't really need it.

bprotopopov replied:
I do not have a preference. I do have a wrapper structure in ztest I can put zfs_rlock_t in instead of znode_t.

tuxoko commented:
@bprotopopov 
I thought the ztest stuff is independent to others, so what I meant is only rebase ztest patch on to my patch in a separate branch.

bprotopopov commented:
@tuxoko yes the ztest commit is independent to my knowledge.

bprotopopov commented:
Just ran 10 iterations of the ztest command that failed on Centos 6.7 - I cannot reproduce the crash.

behlendorf commented:
@bprotopopov the CentOS 6.7 failure is very likely #4267 which is unrelated.  It's the only ztest failure I see these days and it's fairly reproducible if you let zloop run long enough (a few hours).

bprotopopov commented:
@behlendorf, OK, I can look into addressing @tuxoko's comments above. What is the desired timeline for completing this work?

behlendorf commented:
@bprotopopov to my knowledge it's not blocking anything and is mainly a nice bit of cleanup and code improvement.  It's the kind of thing which can be accepted once all the patches are in good shape and have been signed off on.

behlendorf commented:
@bprotopopov can you please refresh this PR, or better yet open a new one and close this one, with the remaining fixes we need to apply.  Several of these patches are now in ZoL master and the 6513 fix has been merged in to the OpenZFS repo.

bprotopopov commented:
Sure, will look into it

________________________________
From: Brian Behlendorf <notifications@github.com>
Sent: Tuesday, May 17, 2016 1:52:31 PM
To: zfsonlinux/zfs
Cc: Boris Protopopov; Mention
Subject: Re: [zfsonlinux/zfs] issue-4023: partial overwrite of metadata holes results in loss of hole birth info (#4024)


@bprotopopov<https://github.com/bprotopopov> can you please refresh this PR, or better yet open a new one and close this one, with the remaining fixes we need to apply. Several of these patches are now in ZoL master and the 6513 fix has been merged in to the OpenZFS repo.

-
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub<https://github.com/zfsonlinux/zfs/pull/4024#issuecomment-219799273>

bprotopopov commented:
@behlendorf, I can't find a fix for 6513 on the master of 6.0.5 branches; perhaps I am looking in the wrong places.

bprotopopov commented:
@behlendorf, OK, I will open a separate pull request to port 6513 from the openzfs repo.
